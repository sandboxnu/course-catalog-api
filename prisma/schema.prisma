generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_WITH_CONNECTIONS")
}

model Course {
  classAttributes String[]         @map("class_attributes")
  classId         String?          @map("class_id")
  coreqs          Json?
  description     String?
  feeAmount       Int?             @map("fee_amount")
  feeDescription  String?          @map("fee_description")
  host            String?
  id              String           @id
  lastUpdateTime  DateTime?        @default(now()) @map("last_update_time")
  maxCredits      Int?             @map("max_credits")
  minCredits      Int?             @map("min_credits")
  name            String?
  nupath          String[]
  optPrereqsFor   Json?            @map("opt_prereqs_for")
  prereqs         Json?
  prereqsFor      Json?            @map("prereqs_for")
  prettyUrl       String?          @map("pretty_url")
  subject         String?
  termId          String?          @map("term_id")
  url             String?
  FollowedCourse  FollowedCourse[]
  sections        Section[]

  @@unique([classId, termId, subject], name: "uniqueCourseProps", map: "uniqueCourseProps")
  @@map("courses")
}

model Major {
  majorId      String @map("major_id")
  plansOfStudy Json   @map("plans_of_study")
  yearVersion  String @map("year_version")
  spec         Json

  @@id([yearVersion, majorId])
  @@map("majors")
}

model Professor {
  id                String  @id
  name              String
  firstName         String  @map("first_name")
  lastName          String  @map("last_name")
  primaryDepartment String? @map("primary_department")
  primaryRole       String? @map("primary_role")
  phone             String?
  email             String?
  officeRoom        String? @map("office_room")

  @@map("professors")
}

model Section {
  classHash       String?           @map("class_hash")
  classType       String?           @map("class_type")
  crn             String?
  honors          Boolean?
  id              String            @id
  info            String?
  meetings        Json?
  campus          String?
  profs           String[]
  seatsCapacity   Int?              @map("seats_capacity")
  seatsRemaining  Int?              @map("seats_remaining")
  url             String?
  waitCapacity    Int?              @map("wait_capacity")
  waitRemaining   Int?              @map("wait_remaining")
  lastUpdateTime  DateTime?         @default(now()) @map("last_update_time")
  course          Course?           @relation(fields: [classHash], references: [id], onDelete: Cascade)
  FollowedSection FollowedSection[]

  @@map("sections")
}

model Subject {
  abbreviation String @id
  description  String

  @@map("subjects")
}

model FollowedCourse {
  courseHash String @map("course_hash")
  userId     Int    @map("user_id")
  course     Course @relation(fields: [courseHash], references: [id], onDelete: Cascade)
  user       User   @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@id([userId, courseHash])
  @@map("followed_courses")
}

model FollowedSection {
  sectionHash String  @map("section_hash")
  userId      Int     @map("user_id")
  section     Section @relation(fields: [sectionHash], references: [id], onDelete: Cascade)
  user        User    @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@id([userId, sectionHash])
  @@map("followed_sections")
}

model User {
  id               Int               @id @default(autoincrement())
  phoneNumber      String            @unique(map: "users.phone_number_unique") @map("phone_number")
  FollowedCourse   FollowedCourse[]
  followedSections FollowedSection[]

  @@map("users")
}

model TermInfo {
  termId     String @unique(map: "term_ids.term_id_unique") @map("term_id")
  subCollege String @map("sub_college")
  text       String
  active     Boolean @default(true)

  @@map("term_ids")
}
