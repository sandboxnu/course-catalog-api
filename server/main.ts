import graphQlServer from "./graphql/server";
import macros from "../utils/macros";
import notificationServer from "./notifs/server";

const port = 8080;
notificationServer.listen(port, () => {
  console.log("Running twilio notification server on port %s", port);
});

graphQlServer
  .listen()
  .then(({ url }) => {
    macros.log(`ready at ${url}`);
    return;
  })
  .catch((err) => {
    macros.error(`error starting graphql server: ${JSON.stringify(err)}`);
  });
