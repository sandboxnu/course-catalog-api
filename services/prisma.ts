import { PrismaClient } from "@prisma/client";
import macros from "../utils/macros";

let prisma: PrismaClient;
try {
  prisma = new PrismaClient({
    log: ["info", "warn", "error"],
  });
  macros.log("** Created Prisma client");

  // TEMP / TODO - REMOVE / DO NOT LEAVE HERE PLEASE
  // Temp fix to address Prisma connection pool issues
  // https://github.com/prisma/prisma/issues/7249#issuecomment-1059719644
  const intervalTime = 6 * 60 * 60_000; // Every 6 hours
  setInterval(async () => {
    const startTime = Date.now();
    await prisma.$disconnect();
    macros.log("Disconnected Prisma");
    await prisma.$connect();
    const totalTime = Date.now() - startTime;
    macros.log(
      `Reconnected Prisma - downtime of ${totalTime} ms (${
        totalTime / 60_000
      } mins)`
    );
  }, intervalTime);
} catch (e) {
  macros.error(e);
}

export default prisma;
