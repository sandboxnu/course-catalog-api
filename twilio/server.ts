import express from "express";
import cors from "cors";
import { createServer } from "http";
import jwt from "jsonwebtoken";
import {
  sendVerificationCode,
  checkVerificationCode,
  handleUserReply,
} from "./notifs";

const corsOptions = {
  origin: process.env.CLIENT_ORIGIN,
};

const app = express();
app.use(cors(corsOptions));
const port = 8080;
app.use(express.json());
const server = createServer(app);

server.listen(port, () => {
  console.log("Running twilio notification server on port %s", port);
});

app.get("/knockknock", (req, res) => res.send("Who's there?"));

app.post("/twilio/sms", (req, res) => handleUserReply(req, res));

app.post("/sms/signup", (req, res) => {
  // twilio needs the phone number in E.164 format see https://www.twilio.com/docs/verify/api/verification
  const phoneNumber = req.body.phoneNumber;
  if (!phoneNumber) {
    res.status(400).send("Missing phone number.");
  }
  sendVerificationCode(phoneNumber)
    .then((response) => {
      res.status(response.statusCode).send(response.message);
    })
    .catch((e) =>
      res.status(500).send("Error trying to send verification code")
    );
});

app.post("/sms/verify", (req, res) => {
  const phoneNumber = req.body.phoneNumber;
  const verificationCode = req.body.verificationCode;
  if (!phoneNumber || !verificationCode) {
    return res.status(400).send("Missing phone number or verification code.");
  }

  checkVerificationCode(phoneNumber, verificationCode)
    .then((response) => {
      if (response.statusCode === 200) {
        const token = jwt.sign({ phoneNumber }, process.env.JWT_SECRET);
        res
          .status(response.statusCode)
          .send({ message: response.message, token });
      } else {
        res.status(response.statusCode).send(response.message);
      }
    })
    .catch((e) => res.status(500).send("Error trying to verify code"));
});
