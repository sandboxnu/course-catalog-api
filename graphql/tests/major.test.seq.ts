import { createTestClient } from "apollo-server-testing";
import { gql } from "apollo-server";
import prisma from "../../services/prisma";
import server from "../index";

const { query } = createTestClient(server);

beforeAll(async () => {
  await prisma.major.deleteMany({});

  await prisma.major.create({
    data: {
      majorId: "computer-information-science/computer-science/bscs",
      yearVersion: "2018",
      spec: { name: "Computer Science, BSCS", yearVersion: 2018 },
      plansOfStudy: [{ years: [1000], id: "0" }],
    },
  });

  await prisma.major.create({
    data: {
      majorId: "computer-information-science/computer-science/bscs",
      yearVersion: "2017",
      spec: { name: "Computer Science, BSCS", yearVersion: 2017 },
      plansOfStudy: [{ years: [1000], id: "0" }],
    },
  });

  await prisma.major.create({
    data: {
      majorId: "science/biochemistry/biochemistry-bs",
      yearVersion: "2018",
      spec: { name: "Biochemistry, BS", yearVersion: 2018 },
      plansOfStudy: [{ years: [1000], id: "0" }],
    },
  });
});

it("gets major from majorId", async () => {
  const res = await query({
    query: gql`
      query major {
        major(majorId: "computer-information-science/computer-science/bscs") {
          majorId
        }
      }
    `,
  });
  expect(res).toMatchSnapshot();
});

it("gets specific occurrence", async () => {
  const res = await query({
    query: gql`
      query major {
        major(majorId: "computer-information-science/computer-science/bscs") {
          majorId
          occurrence(year: 2017) {
            yearVersion
            spec
            plansOfStudy
          }
        }
      }
    `,
  });

  expect(res).toMatchSnapshot();
});

it("gets latest occurrence", async () => {
  const res = await query({
    query: gql`
      query major {
        major(majorId: "computer-information-science/computer-science/bscs") {
          majorId
          latestOccurrence {
            yearVersion
            spec
          }
        }
      }
    `,
  });
  expect(res).toMatchSnapshot();
});

it("cannot find major from non-present majorId", async () => {
  const res = await query({
    query: gql`
      query major {
        major(majorId: "humanities/lovecraftian-studies/lovecraft-studies-ba") {
          majorId
          latestOccurrence {
            yearVersion
          }
        }
      }
    `,
  });
  expect(res).toMatchSnapshot();
});

it("cannot find majorOccurrence from non-present year", async () => {
  const res = await query({
    query: gql`
      query major {
        major(majorId: "computer-information-science/computer-science/bscs") {
          majorId
          occurrence(year: 1984) {
            spec
          }
        }
      }
    `,
  });
  expect(res).toMatchSnapshot();
});
