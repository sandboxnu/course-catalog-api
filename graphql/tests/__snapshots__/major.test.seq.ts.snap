// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cannot find major from non-present majorId 1`] = `
{
  "data": {
    "major": null,
  },
  "errors": [
    [GraphQLError: major not found!],
  ],
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`cannot find majorOccurrence from non-present year 1`] = `
{
  "data": {
    "major": {
      "majorId": "computer-information-science/computer-science/bscs",
      "occurrence": null,
    },
  },
  "errors": [
    [GraphQLError: occurrence not found!],
  ],
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`gets latest occurrence 1`] = `
{
  "data": {
    "major": {
      "latestOccurrence": {
        "spec": {
          "name": "Computer Science, BSCS",
          "yearVersion": 2018,
        },
        "yearVersion": "2018",
      },
      "majorId": "computer-information-science/computer-science/bscs",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`gets major from majorId 1`] = `
{
  "data": {
    "major": {
      "majorId": "computer-information-science/computer-science/bscs",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`gets specific occurrence 1`] = `
{
  "data": {
    "major": {
      "majorId": "computer-information-science/computer-science/bscs",
      "occurrence": {
        "plansOfStudy": [
          {
            "id": "0",
            "years": [
              1000,
            ],
          },
        ],
        "spec": {
          "name": "Computer Science, BSCS",
          "yearVersion": 2017,
        },
        "yearVersion": "2017",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
